keras/models.py

```diff
@@ -806,49 +777,59 @@ def load_weights(self, filepath):
+        # Returns
+            int, number of valid tuple positions
+        '''
+        if not hasattr(generator_output, '__len__'):
+            raise Exception('The generator output must be a tuple.')
+        if len(generator_output) == 2:
+            return 2
+        elif len(generator_output) > 2:
+            return 3
+        else:
+            raise Exception('The generator output tuple must have '
+                            '2, 3, or more elements.')
+
+    def fit_on_generator(self, data_generator, nb_epoch, samples_per_epoch=None,
+                         verbose=1, show_accuracy=False, callbacks=[],
+                         validation_generator=None, class_weight=None):
         '''Fit a model on data generated batch-by-batch by a Python generator.
-        The generator is run in parallel to the model, for efficiency,
-        and can be run by multiple workers at the same time.
-        For instance, this allows you to do real-time data augmentation
-        on images on CPU in parallel to training your model on GPU.
 
         # Arguments
-            generator: a Python generator,
-                yielding either (X, y) or (X, y, sample_weight).
-                The generator is expected to loop over its data
-                indefinitely. An epoch finishes when `samples_per_epoch`
-                samples have been seen by the model.
-                The output of the generator must be a tuple of either 2 or 3
-                numpy arrays.
+
+            data_generator: a Python generator of generators,
+                `data_generator` is supposed to yield `epoch_generator`
+                `epoch_generator` is supposed to yield n-tuple
+                yielding either (X, y) or (X, y, sample_weight) or (X, y, sample_weight, ...).
+                If you have just `epoch_generator`, `data_generator` can be simply created as 
+                `iter(lambda: epoch_generator, None)` which would create infinite generator of the same `epoch_generator`
+                If you have epoch_generator, which needs to be reinitialized each epoch (e.g. ParallelizeIt) you can use
+                `keras.utils.io_utils.generate_infinit_iterators(epoch_generator_constructor, epoch_generator_constructor_args)` 
+                which would create for each epoch new instance
+                X, y and sample_weight should be numpy array. len(y) and len(sample_weight) has to be the same
                 If the output tuple has two elements, they are assumed to be
                 (input_data, target_data).
                 If it has three elements, they are assumed to be
                 (input_data, target_data, sample_weight).
-                All arrays should contain the same number of samples.
+                If there are more than 3 items in tuple (e.g. metadata for scoring, names, etc.),they are ignored for training.
+
             samples_per_epoch: integer, number of samples to process before
```

Comment:
It should be noted that this is an optional epoch break point if your `data/epoch_generator` does not stop.